reset.c

```diff
@@ -0,0 +1,888 @@
+      -r <file path>        aux tags to be rejected in csv file, optional\n\
+      --reject-PG [<file path>] pg names to be rejected in csv file, optional\n\
+      --no-RG               to have RG lines or not, optional\n");
+
+    sam_global_opt_help(fp, "--.--.--");
+    return;
+}
+
+/// getdefaultauxtags - add default aux tags to the collection
+/** @param config pointer to conf_data which holds the aux_set pointer; if the aux_set exists, clear and create new one
+returns 1 on failure and 0 on success
+*/
+int getdefaultauxtags(conf_data *config)
+{
+    int ret = 0;
+
+    if (!config) {
+        return 1;
+    }
+
+    if (config->auxtagsin) {      //free if already exists
+        kh_destroy(aux_set, config->auxtagsin);
+        config->auxtagsin = NULL;
+    }
+
+    config->auxtagsin = kh_init(aux_set);    //initialize the new one
+
+    //add defaults, they may get overridden by reject option
+    kh_put(aux_set, config->auxtagsin, TAGNUM("BC"), &ret);   //barcode
+    kh_put(aux_set, config->auxtagsin, TAGNUM("CB"), &ret);   //cell identifier
+    //TODO add tags which are kept for backward compatibility?
+    kh_put(aux_set, config->auxtagsin, TAGNUM("LB"), &ret);   //library
+    kh_put(aux_set, config->auxtagsin, TAGNUM("MI"), &ret);   //molecular identifier
+
+    return 0;
+}
+
+/// getcsvauxtags - aux tag parsed from csv input file, validate and add to internal set
+/** @param field - the aux tag parsed
+ *  @param data - pointer to conf_data which holds internal flags and aux_set
+returns -1 on failure and 0 on success
+*/
+int getcsvauxtags(const char *field, void *data)
+{
+    int ret = 0, len = 0;
+    conf_data *config = (conf_data*)data;
+
+    if (!field || !config || !config->auxtagsin)
+        return -1;
+
+    len = strlen(field);
+    if (len != 2)                                                       //invalid tag!
+        return -1;
+
+    if (!config->tmpfield)                                              //tmpfield says add or remove
+        kh_put(aux_set, config->auxtagsin, TAGNUM(field), &ret);        //add to set
+    else {                                                              //remove from set
+            khint_t iter = kh_get(aux_set, config->auxtagsin, TAGNUM(field));
+            iter == kh_end(config->auxtagsin) ? 0 : kh_del(aux_set, config->auxtagsin, iter);
+    }
+    return ret < 0 ? -1 : 0;
+}
+
+/// parsecsvfile - parses the csv file and invokes the consumer function / callback
+/** @param field - the aux tag parsed
+ *  @param data - pointer to conf_data which holds internal flags and aux_set
+returns -1 on failure and 0 on success
+data expected in csv format terminated by '\n', parses with ',', callback is invoked to process the tokens
+rolls the buffer incase last char is not '\n' or ','. max supported tag is ~1023 long.
+*/
+int parsecsvfile(FILE *fp, int (*processdata)(const char *, void *), void *data)
```

Comment:
The comments above have `@param field`, but we have `processdata` instead.  I'm guessing this was from an earlier iteration.

Also, you may wish to consider looking into kgetline.  It returns a kstring of the whole line read and it grows automatically to fit it in.  This avoids the whole `partialread` shenanigans when using `fgets`, so it ought to make the code much simpler.

Reply:
You may wish to look at how `samtools view --keep-tag` and `--remove-tag` options work.  We should probably use the same long and short option names too.

Samtools view usage:

```
  -x, --remove-tag STR
               Comma-separated read tags to strip (repeatable) [null]
      --keep-tag STR
               Comma-separated read tags to preserve (repeatable) [null].
               Equivalent to "-x ^STR"
```

The "^" bit is a negation operation.  (I think I copied that nomenclature from some of Petr's bcftools commands.)

You can copy the text out of the doc/samtools-view.1 man page as well.

Reply:
A further oddity about this is that it needs a newline to succeed.  If the last line isn't complete it is silently ignored.  Eg compare:

```
@ seq4c[samtools.../samtools]; ./samtools reset --no-RG -k <(perl -e 'print "NM\n"') -i ~/scratch/data/9827_2#49.1m.bam -o -|grep -c NM:i
99890
@ seq4c[samtools.../samtools]; ./samtools reset --no-RG -k <(perl -e 'print "NM"') -i ~/scratch/data/9827_2#49.1m.bam -o -|grep -c NM:i
0
```

I still think we should just copy the code from samtools view:  https://github.com/samtools/samtools/blob/develop/sam_view.c#L396-L421.  We can move that to utils.c or similar and share the same parsding and hash table functionality.

Reply:
Modified to use command line aux tags as in view instead of csv files. Defaults are not in use.